plonky2/examples/factorial.rs

```diff
@@ -0,0 +1,40 @@
+use anyhow::Result;
+use plonky2::field::types::Field;
+use plonky2::iop::witness::{PartialWitness, Witness};
+use plonky2::plonk::circuit_builder::CircuitBuilder;
+use plonky2::plonk::circuit_data::CircuitConfig;
+use plonky2::plonk::config::{GenericConfig, PoseidonGoldilocksConfig};
+
+/// An example of using Plonky2 to prove a statement of the form
+/// "I know n * (n + 1) * ... * (n + 99)".
+/// When n == 1, this is proving knowledge of 100!.
+fn main() -> Result<()> {
+    const D: usize = 2;
+    type C = PoseidonGoldilocksConfig;
+    type F = <C as GenericConfig<D>>::F;
+
+    let config = CircuitConfig::standard_recursion_config();
+    let mut builder = CircuitBuilder::<F, D>::new(config);
+
+    // The arithmetic circuit.
+    let initial = builder.add_virtual_target();
+    let mut cur_target = initial;
+    for i in 2..101 {
+        let i_target = builder.constant(F::from_canonical_u32(i));
+        cur_target = builder.mul(cur_target, i_target);
+    }
+
+    // Public inputs are the initial value (provided below) and the result (which is generated).
+    builder.register_public_input(initial);
+    builder.register_public_input(cur_target);
+
+    let mut pw = PartialWitness::new();
+    pw.set_target(initial, F::ONE);
+
+    let data = builder.build::<C>();
+    let proof = data.prove(pw)?;
+
+    println!("100 factorial (mod |F|) is: {}", proof.public_inputs[1]);
```

Comment:
I think this is out of date? Could call it `100! / (n - 1)!` or just `factorial range`

Reply:
Oh I guess it's still accurate in a way, since we know n=1 in this code. But I think it would be natural to say something like `println!("Factorial starting at {} is {}!", proof.public_inputs[0], proof.public_inputs[1])`, as if we ignore the private witness data and only consider the proof, that's the statement the proof represents.

Reply:
Ah good idea!

Reply:
Ah good idea!